### DISCUSSION TITLE
New version update: No extra deployment setup is required anymore!

### DISCUSSION THREAD

**Discussion by spencerwooo** [spencerwooo/onedrive-vercel-index#215] [Announcements]
Finally, I managed to land this before 2022. **Happy new year!** 🥳

### TL;DR

For this update, you don't have to create your own client id, client secret, access token, refresh token what-so-ever. The OAuth process is built directly into the project. Tokens are stored with Redis, and refreshed regularly. No more environment variable over the limit issues! 

### Thought process and migration guides

The catch: you need a Redis connection. A url is sufficient, which you would be storing inside Vercel environment variable as `REDIS_URL`. You can create one completely for free at https://upstash.com/

As this upgrade brings a persistent storage into the serverless environment, we can further extract configurations and other settings to this storage, instead of config files - even more simplified deployment process, and even - setting UI or admin panels. Something for me to do in 2022 :)

Migration documentations are available at: [Migrating from an old project](https://github.com/spencerwooo/onedrive-vercel-index#migrating-from-an-old-project)

### Auth process screenshots

![image](https://user-images.githubusercontent.com/32114380/147827921-becd5b4e-ca35-4b4c-82f4-d80777894a0e.png)

![image](https://user-images.githubusercontent.com/32114380/147827959-977ffa5c-afd9-41bd-b41a-b1b5e99f590c.png)

![image](https://user-images.githubusercontent.com/32114380/147828051-c6037e76-657e-48cd-91fd-bc16536b8b5d.png)

**Comment by opoet** [comment]
你好，请问混淆 secret 是啥意思，应该如何填写？

**Reply by spencerwooo** [reply]
client id 和 client secret 都不用动，保持和主项目一致即可。

**Reply by spencerwooo** [reply]
Check out: https://github.com/spencerwooo/onedrive-vercel-index/blob/main/README.md#migrating-from-an-old-project

> You need to revert some fields of your customised config/api.json back, where specifically the clientId and obfuscatedClientSecret should exactly match:
>
> ```json
> {
>  "clientId": "d87bcc39-1750-4ca0-ad54-f8d0efbb2735",
>  "obfuscatedClientSecret": "U2FsdGVkX1830zo3/pFDqaBCVBb37iLw3WnBDWGF9GIB2f4apzv0roemp8Y+iIxI3Ih5ecyukqELQEGzZlYiWg==",
> }
> ```

**Reply by opoet** [reply]
明白了，谢谢 ！

**Reply by comicchang** [reply]
> 明白了，谢谢 ！

如果要用自己的clientId和secret，把secret用aes加密一下就可以了

https://github.com/comicchang/onedrive-vercel-index/blob/38692be104a62f9ca1f4c0ece225c4b3aa676f60/utils/oAuthHandler.ts#L8

**Reply by spencerwooo** [reply]
更新的初衷就是不用自己配置 client id 和 client secret 了……不过道理是一样的，问题不大

**Reply by comicchang** [reply]
deploy当前的head版本，报错 `Task timed out after 10.01 second`，请问有什么建议吗

使用自己的clientId/secret和自带的clientId/secret都是同样的错误

**Reply by spencerwooo** [reply]
是 Vercel 的 serverless function timeout 了看起来，具体是哪个 function 路径？

**Reply by comicchang** [reply]
hmm，大概是`api`？

```
[GET] /api?path=/
12:30:49:04
Function Status:
None
Edge Status:
504
Duration:
10010.63 ms
Init Duration:
N/A
Memory Used:
36 MB
ID:
6hpdh-1641011449030-6651da35c18b
User Agent:
x
2022-01-01T04:30:59.179Z 1c64efc8-7ca3-4514-a148-c1f23da4c555 Task timed out after 10.01 seconds
```

**Reply by spencerwooo** [reply]
这是获取根目录的时候超时了啊，那是去拿 access token 超时，还是 OneDrive API 那侧超时了？要是本地 npm run dev 还能跑起来的话，看看 log？具体在这里 https://github.com/spencerwooo/onedrive-vercel-index/blob/main/pages/api/index.ts#L24-L65 有 access token 相关的 log 获取输出。

**Reply by comicchang** [reply]
在本地调试过后，我根据日志提示搞定了……原来是我的REDIS_URL格式不对

原先写的是`redis://xxx@global-included-jay-31183.upstash.io:31183`，改成`rediss://:xxx@global-included-jay-31183.upstash.io:31183`之后就可以了

谢谢

**Reply by sobuj53** [reply]
Hello, 
I'm also having difficulty understanding how to insert `obfuscatedClientSecret`. Should I left it blank? If I leave it blank then the token doesn't work. I tried to use this online AES converter `https://aesencryption.net/` but no success, I also tried to use bare Client Secret without encryption but this didn't work for me as well. I'm always getting,
`AADSTS7000218: The request body must contain the following parameter: 'client_assertion' or 'client_secret'.`

Can you please tell me what I'm missing? 

PS I've Upstash integration with my vercel account as per the instructions.

**Reply by spencerwooo** [reply]
@sobuj53 if you are using onedrive international with an admin account (meaning that you can authorise `onedrive-vercel-index` to access your account), you don't have to customise anything here, keep the client id and obfuscated client secret the same as the main project. the obfuscation is done here if you must use your own client id and secrets: https://github.com/spencerwooo/onedrive-vercel-index/blob/main/utils/oAuthHandler.ts#L9-L13

**Reply by sobuj53** [reply]
Thank you for your response. I've solved the issue. My account was not accepting the default `clientId` and `obfuscatedClientSecret` and was getting same response as [215](https://github.com/spencerwooo/onedrive-vercel-index/discussions/215#discussioncomment-1894762). However after Obfuscating my generated `ClientSecret` and encrypting it with `onedrive-vercel-index` key. I managed to overcome the issue.

**Comment by hu3rror** [comment]
已经迁移成功。关于 [upstash](https://upstash.com/) 的 `REDIS` 的操作，我新建的时候 REDIS 开了 TLS，所以我 environment 中 `REDIS_URL`需要以 `rediss://:` 开头，如果连接不成功可以试试。

**Comment by tkfanswind** [comment]
这个自动生产是自己的client ID吗？需要换成自己的client ID吗？

**Reply by spencerwooo** [reply]
上面已经提到了。是的，不必替换为原先的自己的 client id 或者 client secret，用项目提供的即可。

**Comment by Kasper4649** [comment]
用的教育版 OneDrive，没有添加未授权应用的权限，只能自己手动添加 Client Id 和 Secret 🌝

**Reply by comicchang** [reply]
https://github.com/spencerwooo/onedrive-vercel-index/discussions/215#discussioncomment-1892784

把你的secret用aes加密一下，填到obfuscatedClientSecret里面

**Comment by coldpub** [comment]
Hello, the third step does not need to change the API, when I click Store tokens, the homepage has been 404

**Reply by coldpub** [reply]
Oops, that's a four-oh-four.
Request failed with status code 500
Press F12 and open devtools for more details, or seek help at onedrive-vercel-index discussions.

**Reply by ChirmyRam** [reply]
> Oops, that's a four-oh-four. Request failed with status code 500 Press F12 and open devtools for more details, or seek help at onedrive-vercel-index discussions.

你好，请问是怎么解决的呢？

**Reply by coldpub** [reply]
我就按照流程来的 最后一步需要等会才会生效 然后强制刷新一下就可以了

**Reply by ChirmyRam** [reply]
> 我就按照流程来的 最后一步需要等会才会生效 然后强制刷新一下就可以了

找到问题了，我redis url少了个冒号🥴

**Reply by coldpub** [reply]
ok!

**Comment by coldpub** [comment]
i ok!

**Comment by jhonnyminedr** [comment]
Hey guys, I did the new way without replacing given cliend id, obsucuted client secret it says admin permissions needed, can I know how to do with my credentials, why is osucutated client secret (any one with problem of admin permissions needed?) at vercel link

in step 2/3 link i got 
![Screenshot (145)](https://user-images.githubusercontent.com/95773407/147864240-a951eb13-47e3-4617-9b99-d9bc495a59de.jpg)
& later i replaced cliend id in this photo
![Screenshot (144)](https://user-images.githubusercontent.com/95773407/147864255-e6f5428b-84eb-4b20-8b4c-c5005853ff8d.png)
got the authorisation code  but after it in later step it says 

![Screenshot (146)](https://user-images.githubusercontent.com/95773407/147864321-4c10e7e8-7977-4039-990b-4e7fc17d4b27.png)

<code>
Step 3/3: Get access and refresh tokens

Whoops, looks like we got a problem: invalid_client.

AADSTS7000215: Invalid client secret is provided.
Trace ID: 8ffc91f6-f44b-44ee-a3fb-7cc360c89e01
Correlation ID: 8c777adb-ff8c-4456-bf87-2193e2494870
Timestamp: 2022-01-02 02:16:37Z

Check out Microsoft's official explanation on the error message.
</code>
i haven't replaced the obsucated client secret can i know how to replace it..
thanks :)

**Reply by spencerwooo** [reply]
the obfuscation is done here https://github.com/spencerwooo/onedrive-vercel-index/blob/main/utils/oAuthHandler.ts#L9-L13

basically we use a default aes encryption with a default encryption key `onedrive-vercel-index`. you can run this function to generate your own encrypted client secret, or use some online tool for this (try googling online aes encryption or something)

**Comment by avgnerd** [comment]
Is there a way to use it with multiple OneDrive accounts?

**Reply by spencerwooo** [reply]
maybe, future versions. for now you can deploy multiple instances of the project (but you'll need multiple redis connections, where the free upstash only provides one)

**Reply by ChirmyRam** [reply]
> maybe, future versions. for now you can deploy multiple instances of the project (but you'll need multiple redis connections, where the free upstash only provides one)

There is no need to create multiple accounts, I found that one account in upstash can create multiple teams, each team has a redis connection, I did just that, now I have created six teams, but the maximum number is unknown.
